hitbox/src/fsm/future.rs

```diff
@@ -279,24 +170,133 @@ where
+                        }
+                        _ => {
+                            ctx.record_state("PollUpstream");
+                            let upstream_future = Box::pin(this.upstream.call(request));
+                            State::PollUpstream {
+                                upstream_future,
+                                permit: None,
+                                ctx: Some(ctx),
+                            }
+                        }
+                    }
+                }
+                StateProj::ConcurrentPollUpstream {
+                    request,
+                    concurrency,
+                    ctx,
+                } => {
+                    let request = request.take().expect(POLL_AFTER_READY_ERROR);
+                    let concurrency = *concurrency;
+                    let mut ctx = ctx.take().expect(CONTEXT_TAKEN_ERROR);
+                    let cache_key = this
+                        .cache_key
+                        .as_ref()
+                        .expect("CacheKey not found for concurrency check");
+                    match this.concurrency_manager.check(cache_key, concurrency) {
+                        ConcurrencyDecision::Proceed(permit) => {
+                            ctx.record_state("PollUpstream");
+                            let upstream_future = Box::pin(this.upstream.call(request));
                             State::PollUpstream {
                                 upstream_future,
+                                permit: Some(permit),
+                                ctx: Some(ctx),
+                            }
+                        }
+                        ConcurrencyDecision::Await(await_future) => {
+                            ctx.record_state("AwaitResponse");
+                            State::AwaitResponse {
+                                await_response_future: await_future,
+                                request: Some(request),
+                                ctx: Some(ctx),
+                            }
+                        }
+                    }
+                }
+                StateProj::AwaitResponse {
+                    await_response_future,
+                    request,
+                    ctx,
+                } => {
+                    let result = ready!(await_response_future.poll(cx));
+                    match result {
+                        Ok(response) => {
+                            // Successfully received response from concurrent request
+                            let mut ctx = ctx.take();
+                            if let Some(ref mut cache_ctx) = ctx {
+                                cache_ctx.record_state("Response");
+                            }
+                            State::Response {
+                                response: Some(response),
+                                ctx,
+                            }
+                        }
+                        Err(concurrency_error) => {
```

Comment:
Should we really handle concurrency error here? Maybe we can move it into LockManager? Since you don't need any additional information from fsm or don't need mutate fsm state.